thewalrus/quantum/means_and_variances.py

```diff
@@ -278,3 +313,122 @@ def variance_clicks(cov, hbar=2):
             term2 += prob_vac_ij - vac_probs[i] * vac_probs[j]
 
     return term1 + 2 * term2
+
+def _coeff_normal_ordered(m, k):
+    r"""Returns the coefficients giving the expansion of a photon number power in terms of normal ordered power of creation
+    and annihilation operators.
+
+    Args:
+        m (int): power of the photon number operator, :math:`(a^\dagger a)^m `.
+        k (int): power of the normal ordered term, :math:`a^{\dagger i} a^i`.
+
+    Returns:
+        (float): expansion coefficient
+    """
+
+    return sum(
+        [
+            (1 / (factorial(mu) * factorial(k - mu)))
+            * ((-1) ** (k - mu) * (mu ** m))
+            for mu in range(0, k + 1)
+        ]
+    )
+
+
+def photon_number_moment(mu, cov, indices, hbar=2):
+    r"""Calculates the expectation value of product of powers of photon number operators of a Gaussian state.
+    The powers are specified by a dictionary with modes as keys and powers as values.
+
+    The calculation is performed by first writing any power of the photon number as
+
+    :math:`(a^\dagger a)^m = \sum_{k=1}^m c_k a^{\dagger k} a^k`
+
+    where the coefficients :math:`c_i` are provided by the function `_coeff_normal_ordered`.
+
+    Args:
+        mu (array): length-:math:`2N` means vector in xp-ordering.
+        cov (array): :math:`2N\times 2N` covariance matrix in xp-ordering.
+        indices (dictionary): specification of the different modes and their power of their photon number
```

Comment:
after reading the tests, this would be a bit of a pain to implement, and it's also a kind of "backend" type function, so I'm very happy to leave it as it is if we want.

Reply:
I was a bit worried about using dictionaries in the walrus, but there is at least one other place where they used in exactly the same, namely in `state_vector` with post-selection. I think I am happy to keep as it is.